src/realm/object-store/sync/sync_session.cpp

```diff
@@ -381,10 +381,28 @@ std::function<void(util::Optional<app::AppError>)> SyncSession::handle_refresh(s
             }
         }
         else if (error) {
-            // 10 seconds is arbitrary, but it is to not swamp the server
-            std::this_thread::sleep_for(milliseconds(10000));
-            if (session_user) {
-                session_user->refresh_custom_data(handle_refresh(session));
+            if (error->http_status_code && *error->http_status_code == 401 &&
+                error->error_code == realm::app::make_error_code(realm::app::ServiceErrorCode::invalid_session)) {
+                // a 401 response on a refresh request means that the token cannot be refreshed and we should not
+                // retry. this can be because an admin has revoked this user's sessions
+                std::unique_lock<std::mutex> lock(session->m_state_mutex);
+                session->cancel_pending_waits(lock, error->error_code);
+                if (session->m_config.error_handler) {
+                    auto user_facing_error = SyncError(
+                        realm::sync::ProtocolError::permission_denied,
+                        "Unable to refresh the user access token; has this user been disabled by an admin?", true);
```

Comment:
I think the error message is a bit misleading here - the UI for revoking sessions is just that - it revokes all refresh tokens for that user but doesn't disable the user. So if they reauthenticate with their credentials, they should be able to obtain a new valid refresh token and restart sync.

Reply:
Turns out I was actually testing the disable feature and not the revoke feature, but I've updated the tests to include revoke as well which as you pointed out is different from disabling a user because logging in again will issue a new refresh token. I've also updated the message emitted here.